Add hoverlabel.zformat attribute

jdugge opened the pull request:
Resolves #1383

This PR adds the `hoverlabel.zformat` attribute, which controls the formatting of the z-values in hover labels.
The attribute can be set both at the `trace` and `layout` level.
If `hoverlabel.zformat` is missing or set to `undefined`, no formatting will occur, leaving the current behaviour unchanged.

jdugge committed:
Add zhoverformat to heatmap and contour traces

jdugge committed:
Add zhoverformat to heatmap and contour traces

jdugge committed:
Add hoverlabel.zformat attribute

etpinard commented on src/components/fx/attributes.js:
```diff
@@ -50,6 +50,17 @@ module.exports = {
                 '`namelength - 3` characters and add an ellipsis.'
             ].join(' ')
         },
+        zformat: {
```
We can't declare `zformat` in `fx/` as it only makes sense for traces with `z` data array (e.g. `heatmap`, `contour` and `surface`).

etpinard replied:
... we wouldn't want traces w/o a `z` attribute to have a `zformat` under their name on https://plot.ly/javascript/reference/

etpinard replied:
... this has me thinking - maybe my `hoverlabel.zformat` idea of https://github.com/plotly/plotly.js/issues/1383#issuecomment-329229633 was wrong. It might be best to keep the `hoverlabel` attribute container _as is_ that way that set of attributes will be the same for all trace types.

etpinard replied:
So maybe `zhoverformat` is the best option. I hate that it doesn't extend to other trace types very well. But I prefer this over breaking `hoverlabel` symmetry.

Any thoughts on this @alexcjohnson @cldougl :question:

alexcjohnson replied:
> So maybe `zhoverformat` is the best option. I hate that it doesn't extend to other trace types very well. But I prefer this over breaking `hoverlabel` symmetry.

I suppose the only other solution would be to move `hoverlabel` into each trace type `attributes` so it can be different for each one. But that's both a much more invasive change, and you're right there's something nice about keeping that container consistent in structure (though a lot of other containers, like `marker` and `line`, DO change from type to type... so it's not without precedent)

Whatever solution we adopt we should think about how it will extend to `intensity` and `surfacecolor`
https://github.com/plotly/plotly.js/issues/1383#issuecomment-280963522 - `surfacecolorhoverformat` is getting a bit unwieldy but... so is `hoverlabel.surfacecolorformat`

So yeah, I guess `zhoverformat` is sounding better and better 😅

cldougl replied:
+1 for `zhoverformat` and keeping `hoverlabel` symmetry at this point.

etpinard commented on src/components/fx/hover.js:
```diff
@@ -1095,7 +1096,13 @@ function cleanPoint(d, hovermode) {
-    if(d.zLabelVal !== undefined) d.zLabel = String(d.zLabelVal);
+    if(d.zLabelVal !== undefined) {
+        if(d.zformat !== undefined) {
+            d.zLabel = d3.format(d.zformat)(d.zLabelVal).replace(/-/g, constants.MINUS_SIGN);
```
Please reuse [`Axes.tickText`](https://github.com/plotly/plotly.js/blob/master/src/plots/cartesian/axes.js#L1167)

etpinard replied:
Referencing https://github.com/plotly/plotly.js/pull/1965, maybe hover-only `format` attributes could support the notion of _stops_ as well.

jdugge replied:
The problem with using `Axes.tickText` is that heatmaps don't have an appropriate z-axis that could be passed to `Axes.tickText`. Should I create a dummy z-axis and use that, similar to what is being done [here](https://github.com/plotly/plotly.js/blob/f1faa2871dbf182847579166af47d1cc2a983707/src/plots/cartesian/axes.js#L601)?

alexcjohnson replied:
> Should I create a dummy z-axis and use that, similar to what is being done here?

Yes, that's the approach we've used in several places. Actually I think we should do that anyway, even without a custom format, as basing the formatting on the actual range of `z` values in the data (ie formatting the hover text the default way we would on an `x` or `y` axis of the same range) would probably solve this issue in most cases anyway, without asking users to pick their own format.

etpinard commented on test/image/mocks/heatmap_hoverlabel.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "data": [
```
A new mock is _ok_, but we should really test the new behavior under 

- https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/heatmap_test.js#L592
- https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/hover_label_test.js#L1238

etpinard commented on src/components/fx/hover.js:
```diff
@@ -1095,7 +1096,13 @@ function cleanPoint(d, hovermode) {
+        if(d.zformat !== undefined) {
+            d.zLabel = d3.format(d.zformat)(d.zLabelVal).replace(/-/g, constants.MINUS_SIGN);
+        } else {
+            d.zLabel = String(d.zLabelVal);
```
I would prefer adding the new formatting logic in the trace `hoverPoints` modules - e.g. this [one](https://github.com/plotly/plotly.js/blob/f1faa2871dbf182847579166af47d1cc2a983707/src/traces/heatmap/hover.js#L112).

etpinard reviewed (commented):
Thanks for the PR!

jdugge commented:
Thanks for the rapid and thorough feedback!

jdugge committed:
Move property out of hoverlabel, use Axis.tickText for formatting, write tests

jdugge committed:
Remove zformat from hoverlabels layout properties

jdugge commented:
Ready for another review. I've moved the option out of the `hoverlabels` and into the `heatmap` trace options and used the `Axis.tickText` method for formatting the labels.

etpinard commented on src/components/fx/hover.js:
```diff
@@ -1095,7 +1095,9 @@ function cleanPoint(d, hovermode) {
     }
 
-    if(d.zLabelVal !== undefined) d.zLabel = String(d.zLabelVal);
+    if(d.zLabelVal !== undefined && d.zLabel === undefined) { // Traces like heatmaps generate the zLabel in their hoverPoints function
```
Hmm. Too bad we need this. Good catch :baseball:  `cleanPoint` could use a little bit of a refactor. 

Would you mind putting that `//  ` comment about the `if()` statement?

etpinard commented on src/traces/heatmap/hover.js:
```diff
@@ -99,6 +104,16 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode, contour)
     }
 
+    var zLabel;
+    var dummyAx = { // dummy axis for formatting the z value
```
Nicely done.

Creating the dummy axis here isn't optimal, but shouldn't make much of a difference. :+1:

etpinard commented on src/traces/heatmap/hover.js:
```diff
@@ -26,6 +27,10 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode, contour)
         y = cd0.y,
         z = cd0.z,
         zmask = cd0.zmask,
+        zmin = trace.zmin,
```
Don't scope variables if you're only going to use them once. The block below could be rewritten as:

```js
   var dummyAx = {
        type: 'linear',
        range: [trace.zmin, trace.zmax],
        hoverformat: trace.zhoverformat,
        _separators: trace._separators
   };
   var zLabel =  Axes.tickText(dummyAx, zVal, 'hover').text;
```

etpinard commented on test/jasmine/tests/hover_label_test.js:
```diff
@@ -499,6 +499,43 @@ describe('hover info', function() {
             .then(done);
         });
+
+        it('should display correct label content with specified format', function(done) {
```
Looks like the two test cases you added are pretty much equivalent. Pick your favorite and :hocho: the other.

etpinard reviewed (commented):
Looking good. This is almost ready for primetime!

To get this thing merged, you should at the very least add `zhoverformat` to `contour` traces. `surface` would benefit from `zhoverformat` too, but I can live without it for now (`surface` should get a `surfacecolorformat` attribute too!).

jdugge committed:
Add zhoverformat to contour plots

jdugge commented:
I added it to `contour` traces and removed one of the test cases.
I tried adding it to `surface` plots, but the z-hoverformat there can already be set using `scene.zaxis.hoverformat`, and it seems that the `surfacecolor` values are never shown on hovering so a `surfacecolorhoverformat` wouldn't do anything?

etpinard commented:
> but the z-hoverformat there can already be set using scene.zaxis.hoverformat, and it seems that the surfacecolor values are never shown on hovering so a surfacecolorhoverformat wouldn't do anything?

Oh good point. I forgot about that.

All right, looks like your PR is ready for the main repo. PR away :rocket:

jdugge committed:
Merge remote-tracking branch 'origin/hoverlabel.zformat' into hoverlabel.zformat

jdugge committed:
Merge branch 'master' into hoverlabel.zformat

jdugge committed:
Fix formatting

etpinard committed:
fixup choropleth select test

- PR https://github.com/plotly/plotly.js/pull/2099 got merged
  on a branch behind https://github.com/plotly/plotly.js/pull/2081
  which caused the test to fail on master.

etpinard committed:
Merge pull request #2109 from plotly/fixup-choropleth-select-test

Fixup choropleth select test

jdugge committed:
Merge branch 'hoverlabel.zformat' of https://github.com/jdugge/plotly.js into hoverlabel.zformat
